mkosi/__init__.py

```diff
@@ -7794,43 +7813,24 @@ def run_qemu_cmdline(config: MkosiConfig, swtpm_socket: Optional[Path] = None) -
                 "scsi-hd,drive=hd,bootindex=1",
             ]
 
+        swtpm_socket = stack.enter_context(start_swtpm())
```

Comment:
hmm, for someone who is not a python guru what precisely does this do? this is a fake "with" expression, right?  but how does destruction here work? when is control returned to start_swtpm() to destroy the tempory dir?

Reply:
So this block starts with the following:

```
with contextlib.Exitstack() as stack:
```

That creates an exit stack, which you can read up on [here](https://docs.python.org/3/library/contextlib.html). To summarize, whenever `stack.enter_context()` is called, the `__enter__` method of the context manager is called, and whatever that returns is propagated as the return value of `enter_context()`. When using `contextlib.contextmanager`, this means everything up to the first `yield` is executed and whatever is passed to `yield` becomes the return value of `enter_context()`.

Cleanup happens when the `__exit__()` method of the exit stack itself is called, which happens at the end of the `with` block that created the exit stack. It will call the `__exit__()` methods of all the context manager that were pushed onto the stack in reverse order. So when the `with` block ends, control is handed back to `start_swtpm()` as it's `__exit__()` method is called along with those of the other context managers that we pushed onto the exit stack.

Reply:
ah, thanks, got it! Learning something new every day!

Reply:
In other words, `ExitStack` allows items in the context manager chain (stack?) to be applied in separate steps, which also means that some of them can be skipped.